src/fields/Fields.cpp

```diff
@@ -493,27 +506,85 @@ void
 Fields::SetBoundaryCondition (amrex::Vector<amrex::Geometry> const& geom, const int lev,
                               std::string component, const int islice)
 {
-    if (lev == 0) return; // keep lev==0 boundaries zero
     HIPACE_PROFILE("Fields::SetBoundaryCondition()");
-    constexpr int interp_order = 2;
+    if (lev == 0 && m_open_boundary) {
+        // Coarsest level: use Taylor expansion of the Green's function
+        // to get Dirichlet boundary conditions
 
-    const amrex::Real ref_ratio_z = Hipace::GetRefRatio(lev)[2];
-    const amrex::Real islice_coarse = (islice + 0.5_rt) / ref_ratio_z;
-    const amrex::Real rel_z = islice_coarse - static_cast<int>(amrex::Math::floor(islice_coarse));
+        amrex::MultiFab staging_area = getStagingArea(lev);
+        // Open Boundaries only work for lev0 with everything in one box
+        amrex::FArrayBox& staging_area_fab = staging_area[0];
 
-    auto solution_interp = interpolated_field_xyz<interp_order>{
-        getField(lev-1, WhichSlice::This, component),
-        getField(lev-1, WhichSlice::Previous1, component),
-        rel_z, geom[lev-1]};
-    amrex::MultiFab staging_area = getStagingArea(lev);
+        const auto arr_staging_area = staging_area_fab.array();
+        const amrex::Box staging_box = staging_area_fab.box();
 
-    for (amrex::MFIter mfi(staging_area, false); mfi.isValid(); ++mfi)
-    {
-        const auto arr_solution_interp = solution_interp.array(mfi);
-        const auto arr_staging_area = staging_area.array(mfi);
-        const amrex::Box fine_staging_box = staging_area[mfi].box();
+        const amrex::Real poff_x = GetPosOffset(0, geom[lev], staging_box);
+        const amrex::Real poff_y = GetPosOffset(1, geom[lev], staging_box);
+        const amrex::Real dx = geom[lev].CellSize(0);
+        const amrex::Real dy = geom[lev].CellSize(1);
+        // scale factor cancels out for all multipole coefficients except the 0th, for wich it adds
+        // a constant therm to the potential
+        const amrex::Real scale = 3._rt/std::sqrt(
+            pow<2>(geom[lev].ProbLength(0)) + pow<2>(geom[lev].ProbLength(1)));
+        const amrex::Real radius = amrex::min(
+            std::abs(geom[lev].ProbLo(0)), std::abs(geom[lev].ProbHi(0)),
+            std::abs(geom[lev].ProbLo(1)), std::abs(geom[lev].ProbHi(1)));
+        AMREX_ALWAYS_ASSERT_WITH_MESSAGE(radius > 0._rt, "The x=0, y=0 coordinate must be inside"
+            "the simulation box as it is used as the point of expansion for open boundaries");
+        // ignore everything outside of 95% the min radius as the Taylor expansion only converges
+        // outside of a circular patch containing the sources, i.e. the sources can't be further
+        // from the center than the closest boundary as it would be the case in the corners
+        const amrex::Real cutoff_sq = pow<2>(0.95_rt * radius * scale);
```

Comment:
Isn't 95% arbitrary? Should it be e.g. min radius minus 3 cells or so?

Reply:
It’s not about a fixed number of Cells, rather it’s to stay a safe distance away from the convergence radius of the Taylor expansion.